### PR TITLE
Improve test sources implementation

### LINKED ISSUE #18138: Eclipse classpath should be able to use test fixtures in implementation dependencies
Similar to https://github.com/gradle/gradle/issues/14932 I have a project with test-fixtures providing utlility classes for testing with certain classes of this module. While I can define a dependency to this fixture and use the classes in other projects tests (inside the same multi project setup) - it is currently not possible to reuse these fixtures as a "production" dependency in Eclipse.

This might seem a bit strange in the first place, but the point is to provide a project that works as a test-framework compiling different test-fixtures and providing additional convenience features for foreign projects that want to do testing with this framework.
For this projects test framework the fixtures are a production dependencies and not something that is going to be used for testing the module.

### Expected Behavior

Declared test artifacts / test fixtures should be available in a dependent projects main `SourceSet`.

### Current Behavior

Test fixtures are only available in the dependent test `SourceSet`s even though the dependency might be declared as implementation.
Either the eclipse plugin should detect usage of the java-test-fixtures out of the box or a configuration should be available to mark SourceSets as "production" code even though it is called "test<something".

### Context

Bundling test-fixtures to be compiled in a convenience test framework.

While I am not familiar with the current limitations and caveats of the eclipse production and test code handling it might be enough to have a option to treat a specific SourceSet as produktion code even though it might be called "test". This would also help in other usecases where the Eclipse project does not allow to have multiple artifacts created by one project.

I've attached a reproducer that works in Intellij and on the command line, while Eclipse fails to compile it.
[eclipse-test-fixtures.zip](https://github.com/gradle/gradle/files/7067086/eclipse-test-fixtures.zip)

### PR DESCRIPTION
**Pull request by donat**
This PR improves how the test and production sources are separated on the Eclipse classpath:
- project dependencies from test configurations now have the`test=true` classpath attribute
- the test source sets and configurations are now configurable via the `eclipse` plugin DSL
- sources and dependencies defined by the JVM test suite plugin are automatically marked with `test=true`

A more elaborate explanation is available in the [documentation changes](https://github.com/gradle/gradle/pull/19310/files#diff-442a0621480993710c571b2be159eef7d90728d9e1064579aeae62d4cbceffe2).

Fixes #18138.

### CODE REVIEW

**Comment by donat** [comment]
@bot-gradle test QuickFeedbackLinux

**Comment by donat** [comment]
@bot-gradle test QuickFeedbackLinux

**Comment by donat** [comment]
@bot-gradle test QuickFeedbackLinux

**Comment by donat** [comment]
@bot-gradle test QuickFeedbackLinux

**Comment by donat** [comment]
@bot-gradle test QuickFeedbackLinux

**Comment by donat** [comment]
@bot-gradle test QuickFeedbackLinux

**Comment by donat** [comment]
@bot-gradle test QuickFeedbackLinux

**Comment by donat** [comment]
@bot-gradle test QuickFeedbackLinux

**Comment by donat** [comment]
@bot-gradle test QuickFeedbackLinux

**Comment by donat** [comment]
@bot-gradle test QuickFeedbackLinux

**Comment by donat** [comment]
@bot-gradle test QuickFeedbackLinux

**Comment by donat** [comment]
@test rFM

**Comment by donat** [comment]
@bot-gradle test rFM

**Comment by donat** [comment]
@bot-gradle test rFM

**Comment by donat** [comment]
@bot-gradle test this

**Comment by donat** [comment]
@bot-gradle test this

**Comment by donat** [comment]
@bot-gradle test rFN

File: subprojects/ide/src/main/java/org/gradle/plugins/ide/eclipse/model/internal/SourceFoldersCreator.java [hunk 1]
```diff
@@ -217,13 +217,9 @@ private boolean containsAll(FileCollection first, FileCollection second) {
...
     }
 
-    private void addSourceSetAttribute(SourceSet sourceSet, SourceFolder folder) {
-        // Using the test sources feature introduced in Eclipse Photon
-        String name = sourceSet.getName();
-        if (!SourceSet.MAIN_SOURCE_SET_NAME.equals(name)) {
-            if (SourceSet.TEST_SOURCE_SET_NAME.equals(name) || folder.getPath().toLowerCase().contains("test")) {
-                folder.getEntryAttributes().put(EclipsePluginConstants.TEST_SOURCES_ATTRIBUTE_KEY, EclipsePluginConstants.TEST_SOURCES_ATTRIBUTE_VALUE);
-            }
+    private void addSourceSetAttribute(SourceSet sourceSet, SourceFolder folder, Collection<SourceSet> testSourceSets) {
```
Comment (tresat) [inline subprojects/ide/src/main/java/org/gradle/plugins/ide/eclipse/model/internal/SourceFoldersCreator.java:220]: Maybe `addTestSourceSetAttributeIfNeeded()`?  I'm a fan of very descriptive method names.

File: subprojects/ide/src/main/java/org/gradle/plugins/ide/eclipse/model/internal/SourceFoldersCreator.java [hunk 2]
```diff
@@ -107,7 +107,7 @@ public String apply(SourceFolder sourceFolder) {
         return collector.apply(Pair.of(regularSourceFolders, dedupedExternalSourceFolders));
     }
 
-    private List<SourceFolder> projectRelativeFolders(Iterable<SourceSet> sourceSets, Function<File, String> provideRelativePath, File defaultOutputDir) {
+    private List<SourceFolder> projectRelativeFolders(Iterable<SourceSet> sourceSets, Collection<SourceSet> testSourceSets, Function<File, String> provideRelativePath, File defaultOutputDir) {
```
Comment (tresat) [inline subprojects/ide/src/main/java/org/gradle/plugins/ide/eclipse/model/internal/SourceFoldersCreator.java:110]: Maybe `configureProjectRelativeFolders()`?

File: subprojects/ide/src/main/java/org/gradle/plugins/ide/eclipse/model/internal/EclipseDependenciesCreator.java [hunk 3]
```diff
@@ -122,7 +125,7 @@ public void visitProjectDependency(ResolvedArtifactResult artifact, boolean asJa
                 buildDependencies = ((ComponentArtifactMetadata) artifactId).getBuildDependencies();
             }
             // TODO: Add handling for Test-only dependencies once https://github.com/gradle/gradle/pull/9484 is merged
```
Comment (tresat) [inline subprojects/ide/src/main/java/org/gradle/plugins/ide/eclipse/model/internal/EclipseDependenciesCreator.java:127]: Remove TODO?

File: subprojects/ide/src/integTest/groovy/org/gradle/plugins/ide/eclipse/EclipseTestConfigurationsWithExternalDependenciesIntegrationTest.groovy [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+
+        when:
+        run 'eclipse'
+
+        then:
+        assertJarDependencyHasTestAttribute('lib-1.0.jar')
+    }
+
+    @ToBeFixedForConfigurationCache
+    def "dependencies in jvm test suites are marked with test classpath attribute"() {
```
Comment (tresat) [inline subprojects/ide/src/integTest/groovy/org/gradle/plugins/ide/eclipse/EclipseTestConfigurationsWithExternalDependenciesIntegrationTest.groovy:69]: I'd add an additional test for a suite which doesn't add an explicit dependency, but uses one of the `useJUnitJupiter()` methods to add its dependencies, just to be sure.

File: subprojects/ide/src/main/java/org/gradle/plugins/ide/eclipse/EclipsePlugin.java [hunk 5]
```diff
@@ -281,15 +292,89 @@ public void execute(GenerateEclipseClasspath task) {
...
                     }
                 });
+
+                SourceSetContainer sourceSets = project.getExtensions().getByType(JavaPluginExtension.class).getSourceSets();
+                sourceSets.configureEach(new Action<SourceSet>() {
+                    @Override
+                    public void execute(SourceSet sourceSet) {
+                        if (sourceSet.getName().toLowerCase(Locale.ROOT).contains("test")) {
+                            // source sets with 'test' in their name are marked as test on the Eclipse classpath
+                            model.getClasspath().getTestSourceSets().convention(ImmutableSet.<SourceSet>builder()
```
Comment (tresat) [inline subprojects/ide/src/main/java/org/gradle/plugins/ide/eclipse/EclipsePlugin.java:302]: So each time we configure a new SourceSet, we build a new ImmutableSet containing all the existing SourceSets + the new one?  

Wouldn't it be more efficient to set the convention to a local mutable set declared outisde of `configureEach()`, and then have configure callback insert each new sourceset into the mutable set?

Comment (tresat) [reply subprojects/ide/src/main/java/org/gradle/plugins/ide/eclipse/EclipsePlugin.java:302]: Does this pattern not work?  If not, this block of logic to create a new immutable set by adding the new item to the existing set would be better extracted to a diferent private method.

**Review by tresat — APPROVED** [review]
Looks like it covers everything to me.  My suggestions are all minor things.

File: subprojects/ide/src/main/java/org/gradle/plugins/ide/eclipse/model/internal/SourceFoldersCreator.java (hunk 1 above)
Comment (donat) [reply subprojects/ide/src/main/java/org/gradle/plugins/ide/eclipse/model/internal/SourceFoldersCreator.java]: Good idea. Method name adjusted.

File: subprojects/ide/src/main/java/org/gradle/plugins/ide/eclipse/model/internal/EclipseDependenciesCreator.java (hunk 3 above)
Comment (donat) [reply subprojects/ide/src/main/java/org/gradle/plugins/ide/eclipse/model/internal/EclipseDependenciesCreator.java]: This is no longer relevant; removed.

File: subprojects/ide/src/integTest/groovy/org/gradle/plugins/ide/eclipse/EclipseTestConfigurationsWithExternalDependenciesIntegrationTest.groovy (hunk 4 above)
Comment (donat) [reply subprojects/ide/src/integTest/groovy/org/gradle/plugins/ide/eclipse/EclipseTestConfigurationsWithExternalDependenciesIntegrationTest.groovy:69]: Done

File: subprojects/docs/src/docs/userguide/core-plugins/eclipse_plugin.adoc [hunk 6]
```diff
@@ -272,3 +272,45 @@ include::sample[dir="snippets/ide/eclipse/groovy",files="build.gradle[tags=wtp-w
...
+- Project dependencies have the `without_test_sources=false` attribute if the target project applies the java-test-fixtures plugin.
+
+You can customize the test source sets and configurations:
+
+.Customizing test sources
+====
+include::sample[dir="snippets/ide/eclipse/groovy",files="build.gradle[tags=test-sources]"]
+include::sample[dir="snippets/ide/eclipse/kotlin",files="build.gradle.kts[tags=test-sources]"]
+====
+
```
[the remark this answers is not in the document]
Comment (donat) [reply subprojects/docs/src/docs/userguide/core-plugins/eclipse_plugin.adoc:307]: That sounds good to me.

File: subprojects/ide/src/crossVersionTest/groovy/org/gradle/plugins/ide/tooling/r56/ToolingApiEclipseModelTestSourcesCrossVersionSpec.groovy [hunk 7]
```diff
@@ -148,4 +149,100 @@ class ToolingApiEclipseModelTestSourcesCrossVersionSpec extends ToolingApiSpecif
...
+                id 'java-test-fixtures'
+            }
+        """
+
+        when:
+        EclipseProject project = loadToolingModel(EclipseProject)
+        EclipseProject projectA = project.children[0]
+        EclipseProjectDependency depB = projectA.projectDependencies.find { it.path == 'b' }
+        EclipseProjectDependency depC = projectA.projectDependencies.find { it.path == 'c' }
+        EclipseProjectDependency depD = projectA.projectDependencies.find { it.path == 'd' }
```
[the remark this answers is not in the document]
Comment (donat) [reply subprojects/ide/src/crossVersionTest/groovy/org/gradle/plugins/ide/tooling/r56/ToolingApiEclipseModelTestSourcesCrossVersionSpec.groovy:192]: It's a matter of taste I think. There are several tests in the codebase where the references for the asserted objects are gathered in the `when` block (just like here).

File: subprojects/ide/src/crossVersionTest/groovy/org/gradle/plugins/ide/tooling/r56/ToolingApiEclipseModelTestSourcesCrossVersionSpec.groovy [hunk 8]
```diff
@@ -148,4 +149,100 @@ class ToolingApiEclipseModelTestSourcesCrossVersionSpec extends ToolingApiSpecif
...
+        EclipseProject project = loadToolingModel(EclipseProject)
+        def mainSrcDir = project.sourceDirectories.find { it.path == 'src/main/java' }
+        def testSrcDir = project.sourceDirectories.find { it.path == 'src/test/java' }
+        def guava = project.classpath.find { it.file.name.contains 'guava' }
+        def junit = project.classpath.find { it.file.name.contains 'junit' }
+
+        then:
+        !hasTestAttributes(testSrcDir)
+
+        and:
```
[the remark this answers is not in the document]
Comment (donat) [reply subprojects/ide/src/crossVersionTest/groovy/org/gradle/plugins/ide/tooling/r56/ToolingApiEclipseModelTestSourcesCrossVersionSpec.groovy]: I had wrong assumptions about the semantics of the `and:` labels. Removed.

**Comment by donat** [comment]
@bot-gradle test and merge

**Comment by donat** [comment]
@bot-gradle test and merge

**Comment by donat** [comment]
@bot-gradle test and merge

**Comment by blindpirate** [comment]
This seems to have a stable 3% regression in scenario `JavaIDEModelPerformanceTest.get IDE model for IDEA`.

**Comment by donat** [comment]
This makes very little sense as I didn't change any functionality in the idea plugin. I'll check the regression later.

**Comment by donat** [comment]
@bot-gradle test rFN

**Comment by donat** [comment]
@bot-gradle test and merge

**Pull request merged**